lib/study_cleanup_tools.rb

```diff
@@ -0,0 +1,159 @@
+# study_cleanup_tools.rb
+#
+# module for ensuring that each test build starts with clean schemas and Terra billing projects
+# methods are safeguarded so that they cannot be run on deployed instances
+# methods can be run against a local instance by passing allow_dev_env: true (but not recommended)
+
+module StudyCleanupTools
+
+  # disallow running an cleanup commands on deployed production & staging hosts
+  DISALLOWED_HOSTS = /singlecell(-staging)?\.broadinstitute/
+
+  # disallow running deletes against any of these projects
+  DISALLOWED_BILLING_PROJECTS = %w(single-cell-portal single-cell-portal-staging)
+
+  # delete all studies and remove associated Terra workspaces & Google buckets
+  # can only be invoked in the test or pentest environment
+  #
+  # * *params*
+  #   - +allow_dev_env+: (Boolean) => allow execution of this method in development environment (default: false)
+  #
+  # * *raises*
+  #   - (ArgumentError) => if command cannot be run for failing a security check (hostname, project, environment, etc)
+  #
+  # * *returns*
+  #   - (TrueClass) => true if method was performed, otherwise ArgumentError is thrown
+  def self.destroy_all_studies_and_workspaces(allow_dev_env: false)
+    if validate_environment!(allow_dev_env) && validate_hostname!
+      Study.all.each do |study|
+        # do not mass delete studies in protected projects
+        study.destroy_and_remove_workspace if permit_billing_project?(study.firecloud_project)
+      end
+      true
+    end
+  end
+
+  # delete all Terra workspaces in a project that do not have a study associated with them in the given env/schema
+  # will ignore any workspaces created by different service accounts/users
+  # by default will only run against test/pentest environment during continuous integration build
+  #
+  # * *params*
+  #   - +project_name+: (String) => Name of Terra billing project (defaults to FireCloudClient::PORTAL_NAMESPACE )
+  #   - +allow_dev_env+: (Boolean) => allow execution of this method in development environment (default: false)
+  #
+  # * *raises*
+  #   - (ArgumentError) => if command cannot be run for failing a security check (hostname, project, environment, etc)
+  #
+  # * *returns*
+  #   - (TrueClass) => true if method was performed, otherwise ArgumentError is thrown
+  def self.delete_all_orphaned_workspaces(project_name = FireCloudClient::PORTAL_NAMESPACE, allow_dev_env: false)
+    if validate_hostname! && validate_environment!(allow_dev_env) && validate_billing_project!(project_name) && validate_continuous_integration!
```

Comment:
I appreciate all of these guards.

Reply:
I tried to make this as "safe" as possible.  I would be interested to hear if you can think of any other safeguards we can put in place for production/staging data, that would still enable a developer to purge their local environment/billing project if need be.

Reply:
I've actually seen a lot of different explanations for "bang methods" and what they actually mean/do.  While generally it denotes that you're modifying an object in place, I've also read that it simply denotes that the method is "dangerous".  I'm using it here because rather than returning `false` for a validation, it's actually throwing an error.  I did this because I wanted to be _absolutely sure_ that execution is being halted if something doesn't pass one of these checks.  I added the `return true` statements later because I wanted to actually be able to test these, and having a return value made that possible.

On reflection, you're right - it's mixing metaphors, and is generally confusing.  A method called `validate_[something]` in ActiveRecord or another ORM will return true/false rather than explicitly raising an exception.  I think it would probably make a lot more sense for the parent method to raise an exception rather than the validator.  It just seemed DRYer to have the exceptions in the validations.